### PR TITLE
[SE-0279] Add support for an unbraced syntax for multiple trailing closures

### PR DESCRIPTION
**Pull request by rjmccall**
This PR implements the syntax proposed in SE-0279, allowing an initial unlabeled trailing closure to be followed by an arbitrary number of labeled-closure postfixes.

swift-syntax: https://github.com/apple/swift-syntax/pull/212

### CODE REVIEW

**Comment by rjmccall** [comment]
@swift-ci Please smoke test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/212
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/212
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/212
@swift-ci Please test

**Comment by benlangmuir** [comment]
apple/swift-syntax#212
@swift-ci please test

**Comment by benlangmuir** [comment]
apple/swift-syntax#212
@swift-ci please test

File: lib/IDE/CodeCompletion.cpp [hunk 1]
```diff
@@ -976,9 +981,51 @@ void CodeCompletionResultBuilder::addCallParameter(Identifier Name,
...
+          OS << param.getPlainType()->getString(PO);
+          if (param.isVariadic())
+            OS << "...";
+          OS << "#>";
+        }
+      }
+      if (hasSignature)
+        OS << ")";
+      if (!returnsVoid)
+        OS << " -> " << AFT->getResult()->getString(PO);
```
Comment (rintaro) [inline lib/IDE/CodeCompletion.cpp:1015]: @benlangmuir If the return type is unresolved type, could you stop emitting `-> _`?

Comment (benlangmuir) [reply lib/IDE/CodeCompletion.cpp:1015]: Do you have a suggestion for a test case? Everything I've tried either gives a name or it fails entirely to provide argument completions.

**Comment by benlangmuir** [comment]
It appears that CI ran the tests against this branch directly without first merging it into master.  We probably need to rebase.
> GitHub pull request #31052 of commit f1fff73f18470adffe9f5e547310b425edc2a5aa, has merge conflicts.
> swift                              : f1fff73f18470adffe9f5e547310b425edc2a5aa

File: lib/IDE/CodeCompletion.cpp (hunk 1 above)
Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp:1015]: Maybe a generic param type or a archetype type. Sorry I don't have this branch in my hand, but does this suggests `y: { () -> A in <#code#> }`? 
```swift
func foo<A>(x: () -> Void, y: () -> A) {}
foo() {} #^COMPLETE^#
```
If so, I think we should not suggest the signature in this case.

Comment (benlangmuir) [reply lib/IDE/CodeCompletion.cpp:1015]: Ah yep it gives `y: { () -> A in <#code#> }`.  This is consistent with how completing the whole "foo" call behaves though (if you complete foo and then expand the placeholder the result is identical), so if we're going to change it I think it should be a separate PR.

In general, we might want to drop the return type in most cases even if it is resolved - it seems to be more common style to omit the types unless they are required.  We could also drop the parens around the parameters in that case.

Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp:1015]: > I think it should be a separate PR

Agreed!

**Comment by benlangmuir** [comment]
apple/swift-syntax#212
@swift-ci please test

**Comment by benlangmuir** [comment]
apple/swift-syntax#212
@swift-ci please test

**Comment by benlangmuir** [comment]
apple/swift-syntax#212
@swift-ci please test macOS

**Comment by benlangmuir** [comment]
apple/swift-syntax#212
@swift-ci please test

**Comment by rintaro** [comment]
apple/swift-syntax#212
@swift-ci please test

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/216
@swift-ci Please test

**Comment by rintaro** [comment]
apple/swift-syntax#216
@swift-ci Please test

**Comment by xedin** [comment]
@rjmccall Looks like there was a new `CallExpr::create` recently introduced into function builders, you'd have a fix it to use array instead of a single flag.

**Comment by rjmccall** [comment]
apple/swift-syntax#212
@swift-ci please test

**Comment by rintaro** [comment]
@rjmccall I changed swift-syntax PR to apple/swift-syntax#216 because the CI somehow didn't picked up the updated hash of it.

**Comment by rintaro** [comment]
apple/swift-syntax#216
@swift-ci Please test

**Comment by xwu** [comment]
@rjmccall I have updated the PR description to reflect that what's implemented is in fact the proposed syntax and not an alternative to it, for future readers who may be confused.

**Pull request merged**